experiments/scripts/test_pa.py

```diff
@@ -0,0 +1,414 @@
+            "-b", str(config.batch_size),
+            "--concurrency-range", str(config.concurrency),
+            "--max-threads", str(config.max_threads)
+        ]
+        if config.protocol == "http":
+            cmd += ["-u", "localhost:8000"]
+        else:
+            cmd += ["-u", "localhost:8001"]
+
+        if config.is_async:
+            cmd += ["--async"]
+        else:
+            cmd += ["--sync"]
+        # yapf: enable
+        return cmd
+
+    def _get_dict_combos(self, config: dict):
+        param_combinations = list(product(*tuple(config.values())))
+        return [dict(zip(config.keys(), vals)) for vals in param_combinations]
+
+    def _get_configs(self, config: dict):
+        dict_combos = self._get_dict_combos(config)
+        configs = []
+
+        for c in dict_combos:
+            config = RunConfigData()
+            for k, v in c.items():
+                setattr(config, k, v)
+            configs.append(config)
+        return configs
+
+
+class TritonServer():
+    """ Starts and stops a single triton server """
+
+    def __init__(self):
+        self._proc = None
+
+    def start(self, model_repo, model):
+        for proc in psutil.process_iter():
+            try:
+                # Check if process name contains the given name string.
+                if "tritonserver" in proc.name().lower():
+                    raise Exception("Tritonserver already running")
+            except (psutil.NoSuchProcess, psutil.AccessDenied,
+                    psutil.ZombieProcess):
+                pass
+
+        print(
+            f"Starting tritonserver with repo={model_repo}, model={model_name}")
+        cmd = self._get_cmd(model_repo, model)
+        self._proc = self._create_process(cmd)
+        sleep(5)
+        return self._proc.pid
+
+    def stop(self):
+        print(f"Stopping tritonserver")
+        if self._proc is not None:
+            self._proc.terminate()
+            try:
+                self._proc.communicate(timeout=10)
+            except TimeoutExpired:
+                self._proc.kill()
+                self._proc.communicate()
+        else:
+            print(f"TritonServer does not exist?!")
+
+    def _create_process(self, cmd):
+        self._triton_log = tempfile.NamedTemporaryFile()
+        try:
+            process = Popen(cmd,
+                            start_new_session=True,
+                            stdout=self._triton_log,
+                            stderr=STDOUT,
+                            encoding='utf-8')
+        except FileNotFoundError as e:
+            raise Exception(f"command failed immediately: {e}")
+        return process
+
+    def _get_cmd(self, model_repo, model):
+        cmd = [
+            "tritonserver", "--model-repository", model_repo, "--http-port",
+            "8000", "--grpc-port", "8001", "--model-control-mode", "explicit",
+            "--load-model", model
+        ]
+        return cmd
+
+
+nvmlInit()
+# FIXME hardcoded to 1 GPU
```

Comment:
Does this need to be resolved?

Reply:
just trying to keep this script available if it needs to be improved in the future. If we need to do any multi-GPU PA testing then this would have to be fixed then.